### ISSUE TITLE
Extra debug target in win32/makefile.mk

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14340]
Migrated from [rt.perl.org#123439](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123439) (status was 'resolved')

Searchable as RT123439$

**Comment by p5pRT** [comment]
### From @kmx

#### Created by @kmx



win32/makefile\.mk currently supports debug build via&#8203;:

dmake CFG=Debug

which uses&#8203;:
OPTIMIZE = \-g \-O2 \-DDEBUGGING
LINK\_DBG = \-g

I would like to propose adding a special debug target \(e\.g\. DebugSymbols\)
which will not use \-DDEBUGGING
OPTIMIZE = \-g \-O2
LINK\_DBG = \-g

It might be handy if you want a perl binary with debugging symbols but
without a special debugging code which is turned on via \-DDEBUGGING

Thanks

\-\-
kmx

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl 5.20.1:

Configured by strawberry-perl at Mon Sep 15 13:28:28 2014.

Summary of my perl5 (revision 5 version 20 subversion 1) configuration:

   Platform:
     osname=MSWin32, osvers=6.3, archname=MSWin32-x64-multi-thread
     uname='Win32 strawberry-perl 5.20.1.1 #1 Mon Sep 15 13:26:45 2014 x64'
     config_args='undef'
     hint=recommended, useposix=true, d_sigaction=undef
     useithreads=define, usemultiplicity=define
     use64bitint=define, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='gcc', ccflags =' -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE 
-DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS 
-DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields',
     optimize='-s -O2',
     cppflags='-DWIN32'
     ccversion='', gccversion='4.8.3', gccosandvers=''
     intsize=4, longsize=4, ptrsize=8, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='long 
long', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='g++.exe', ldflags ='-s -L"C:\tmp64\perl\lib\CORE" -L"C:\tmp64\c\lib"'
     libpth=C:\tmp64\c\lib C:\tmp64\c\x86_64-w64-mingw32\lib 
C:\tmp64\c\lib\gcc\x86_64-w64-mingw32\4.8.3
     libs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr 
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
     perllibs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr 
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
     libc=, so=dll, useshrplib=true, libperl=libperl520.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_win32.xs, dlext=xs.dll, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-mdll -s -L"C:\tmp64\perl\lib\CORE" 
-L"C:\tmp64\c\lib"'



@INC for perl 5.20.1:
     C:/tmp64/perl/site/lib
     C:/tmp64/perl/vendor/lib
     C:/tmp64/perl/lib
     .


Environment for perl 5.20.1:
     HOME=C:\tmp64\data
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
PATH=C:\tmp64\perl\site\bin;C:\tmp64\perl\bin;C:\tmp64\c\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem
     PERL_BADLANG (unset)
     SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Tue Dec 16 00&#8203;:41&#8203;:51 2014\, kmxx wrote&#8203;:
> This is a bug report for perl from kmx@&#8203;atlas\.cz\,
> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> win32/makefile\.mk currently supports debug build via&#8203;:
> 
> dmake CFG=Debug
> 
> which uses&#8203;:
> OPTIMIZE = \-g \-O2 \-DDEBUGGING
> LINK\_DBG = \-g
> 
> I would like to propose adding a special debug target \(e\.g\.
> DebugSymbols\)
> which will not use \-DDEBUGGING
> OPTIMIZE = \-g \-O2
> LINK\_DBG = \-g
> 
> It might be handy if you want a perl binary with debugging symbols but
> without a special debugging code which is turned on via \-DDEBUGGING
> 

I have no objections to this\. The trickiest thing will be thinking of a name for the new build flavour\.

Presumably you would want it enabled by setting CFG=Something in the makefile\.mk \(and Makefile\)\, but what is the "Something" to be called?

DebugWithoutDebugging?
MinimalDebug?
PlainDebug?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

On Tue Dec 16 00&#8203;:41&#8203;:51 2014\, kmxx wrote&#8203;:
> This is a bug report for perl from kmx@&#8203;atlas\.cz\,
> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> win32/makefile\.mk currently supports debug build via&#8203;:
> 
> dmake CFG=Debug
> 
> which uses&#8203;:
> OPTIMIZE = \-g \-O2 \-DDEBUGGING
> LINK\_DBG = \-g
> 
> I would like to propose adding a special debug target \(e\.g\.
> DebugSymbols\)
> which will not use \-DDEBUGGING
> OPTIMIZE = \-g \-O2
> LINK\_DBG = \-g

Why is it \-O2 being used for CFG=Debug? Wouldn't the \-O2 make using a C debugger hard or impossible to use \(C autos are gibberish and or renamed\, incoming C function args get reused for other things\)? Shouldn't it be "\-O0" for CFG=Debug like it is for Visual C?

I often create \-Od \(VCs \-O0\) non\-DEBUGGING perl when I want a debugger steppable core\, but I dont want to recompile all my XS modules since DEBUGGING status must match between modules and core in practice \(there is a bug open on RT where I state that while others claim that many years ago DEBUGGING was supposed to be ABI compatible with non\-DEBUGGING\, I find that hasn't been in practice for years\) but you want a \-O2 non\-DEBUGGING build with symbols? \("with symbols" doesn't exist for VC Perl since VC Perl build process always makes \.pdb files regardless of optimization level/CFG setting \(PS\, that is not a bad idea\)\)\.

Kmx why dont you make GCC/LD produce separate symbols files like Visual C does https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html and ship the \.debug files in a tarball that is extracted ontop of a SPerl installation? I am guessing this proposal is for http&#8203;://strawberryperl\.com/beta\.html releases right?

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @kmx


On 18\.12\.2014 23&#8203;:37\, bulk88 via RT wrote&#8203;:
> On Tue Dec 16 00&#8203;:41&#8203;:51 2014\, kmxx wrote&#8203;:
>> This is a bug report for perl from kmx@&#8203;atlas\.cz\,
>> generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> win32/makefile\.mk currently supports debug build via&#8203;:
>>
>> dmake CFG=Debug
>>
>> which uses&#8203;:
>> OPTIMIZE = \-g \-O2 \-DDEBUGGING
>> LINK\_DBG = \-g
>>
>> I would like to propose adding a special debug target \(e\.g\.
>> DebugSymbols\)
>> which will not use \-DDEBUGGING
>> OPTIMIZE = \-g \-O2
>> LINK\_DBG = \-g
> Why is it \-O2 being used for CFG=Debug? Wouldn't the \-O2 make using a C debugger hard or impossible to use \(C autos are gibberish and or renamed\, incoming C function args get reused for other things\)? Shouldn't it be "\-O0" for CFG=Debug like it is for Visual C?
>
> I often create \-Od \(VCs \-O0\) non\-DEBUGGING perl when I want a debugger steppable core\, but I dont want to recompile all my XS modules since DEBUGGING status must match between modules and core in practice \(there is a bug open on RT where I state that while others claim that many years ago DEBUGGING was supposed to be ABI compatible with non\-DEBUGGING\, I find that hasn't been in practice for years\) but you want a \-O2 non\-DEBUGGING build with symbols? \("with symbols" doesn't exist for VC Perl since VC Perl build process always makes \.pdb files regardless of optimization level/CFG setting \(PS\, that is not a bad idea\)\)\.

I am fine with whatever changes to CFG=Debug case\. What I want is to build 
with same optimalization as production perl without extra \(non\-production\) 
\-DDEBUGGING code but with debug symbol \(gcc's \-g\)


> Kmx why dont you make GCC/LD produce separate symbols files like Visual C does https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html and ship the \.debug files in a tarball that is extracted ontop of a SPerl installation? I am guessing this proposal is for http&#8203;://strawberryperl\.com/beta\.html releases right?

As you can see here&#8203;:
https://metacpan.org/source/KMX/Perl-Dist-Strawberry-4.004/lib/Perl/Dist/Strawberry/Step/InstallPerlCore.pm#L263
the strawberry build machine is more or less ready for creating separate 
debug info\, but it did not work reliably with current version of gcc / 
binutils \(objcopy from time to time damages some DLLs\)

But yes\, it is the long term goal\.

\-\-
kmx

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Tue Dec 16 01&#8203;:08&#8203;:10 2014\, shay wrote&#8203;:
> On Tue Dec 16 00&#8203;:41&#8203;:51 2014\, kmxx wrote&#8203;:
> > This is a bug report for perl from kmx@&#8203;atlas\.cz\,
> > generated with the help of perlbug 1\.40 running under perl 5\.20\.1\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > win32/makefile\.mk currently supports debug build via&#8203;:
> >
> > dmake CFG=Debug
> >
> > which uses&#8203;:
> > OPTIMIZE = \-g \-O2 \-DDEBUGGING
> > LINK\_DBG = \-g
> >
> > I would like to propose adding a special debug target \(e\.g\.
> > DebugSymbols\)
> > which will not use \-DDEBUGGING
> > OPTIMIZE = \-g \-O2
> > LINK\_DBG = \-g
> >
> > It might be handy if you want a perl binary with debugging symbols
> > but
> > without a special debugging code which is turned on via \-DDEBUGGING
> >
> 
> I have no objections to this\. The trickiest thing will be thinking of
> a name for the new build flavour\.
> 
> Presumably you would want it enabled by setting CFG=Something in the
> makefile\.mk \(and Makefile\)\, but what is the "Something" to be called?
> 
> DebugWithoutDebugging?
> MinimalDebug?
> PlainDebug?


Thanks for the suggestion\. I've now done this in commit 31c916d28f\, using your original suggestion of DebugSymbols for the new CFG option name\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
